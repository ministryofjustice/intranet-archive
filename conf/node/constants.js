import { parseSchduleString } from "./controllers/schedule.js";

export const ordinalNumber = process.env.ORDINAL_NUMBER;
export const port = 2000;
export const jwt = process.env.JWT;

/**
 * Access
 */

export const sharedSecret = process.env.INTRANET_ARCHIVE_SHARED_SECRET;

/**
 * S3
 */

export const s3Region = "eu-west-2";
export const s3BucketName = process.env.S3_BUCKET_NAME;
export const s3Credentials = process.env.AWS_ACCESS_KEY_ID &&
  process.env.AWS_SECRET_ACCESS_KEY && {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
  };

/**
 * CloudFront
 */

export const cloudFrontKeysObject =
  process.env.AWS_CLOUDFRONT_PUBLIC_KEYS_OBJECT;
export const cloudFrontPublicKey = process.env.AWS_CLOUDFRONT_PUBLIC_KEY;
export const cloudFrontPrivateKey = process.env.AWS_CLOUDFRONT_PRIVATE_KEY;

/**
 * Options
 */

export const corsOptions = {
  methods: ["POST"],
  origin: [
    "http://spider.intranet.docker/",
    "https://dev-intranet-archive.apps.live.cloud-platform.service.justice.gov.uk/",
  ],
};

/**
 * Defaults
 */

export const defaultAgency = "hq";
export const defaultUrl = "https://intranet.justice.gov.uk/";

/**
 * Validation
 */

export const allowedTargetHosts = [
  "intranet.docker",
  "intranet.justice.gov.uk",
  "dev.intranet.justice.gov.uk",
  "staging.intranet.justice.gov.uk",
  "demo.intranet.justice.gov.uk",
];

export const allowedTargetAgencies =
  process.env.ALLOWED_AGENCIES?.split(",") ?? [];

/**
 * Schedule
 */

export const snapshotSchedule = parseSchduleString(
  process.env.SNAPSHOT_SCHEDULE,
);

/**
 * Httrack
 */

export const sensitiveFiles = [
  "cookies.txt", // Contains the JWT and CloudFront cookies.
  "hts-in_progress.lock", // Contains the JWT.
  "hts-log.txt", // Has the httrack command line arguments - this includes the JWT.
  `hts-cache/doit.log`, // Has the httrack command line arguments - this includes the JWT.
  `hts-cache/new.zip`,
];

/**
 * Intranet application
 */

export const heartbeatEndpoint = "auth/heartbeat";

/**
 * Index pages
 */

export const indexCss = `
  main{display:block}a{background-color:transparent;color:#337ab7;text-decoration:none;text-shadow:#fff 1px 0 10px;}
  .list-group-item,body{background-color:#fff}a:active,a:hover{outline:0}h1{font-size:2em;margin:.67em 0}
  *,:after,:before{-webkit-box-sizing:border-box;-moz-box-sizing:border-box;box-sizing:border-box}
  html{font-family:sans-serif;-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;font-size:10px;-webkit-tap-highlight-color:transparent}
  body{margin:0;font-family:"Helvetica Neue",Helvetica,Arial,sans-serif;font-size:14px;line-height:1.42857143;color:#333}
  a:focus,a:hover{color:#23527c;text-decoration:underline}
  a:focus{outline:-webkit-focus-ring-color auto 5px;outline-offset:-2px}
  .container{padding-right:15px;padding-left:15px;margin-right:auto;margin-left:auto}
  @media (min-width:768px){.container{width:750px}}@media (min-width:992px){.container{width:970px}}
  @media (min-width:1200px){.container{width:1170px}}
  .list-group{padding-left:0;margin-bottom:20px;display:grid;grid-template-columns:repeat(auto-fill,minmax(160px,1fr))}
  .list-group-item{position:relative;display:block;padding:14px 18px;margin:0 20px 25px 0;background: #fff;
    -webkit-box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
    -moz-box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
    box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
  }
  .container:after,.container:before{display:table;content:" "}
  .container:after{clear:both}.list-group{-webkit-box-shadow:0 1px 2px rgba(0,0,0,.075);box-shadow:0 1px 2px rgba(0,0,0,.075)}
`;
