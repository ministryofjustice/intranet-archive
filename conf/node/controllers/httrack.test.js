import fs from "node:fs";
import { afterAll, beforeEach, it, jest } from "@jest/globals";

import {
  getHttrackArgs,
  runHttrack,
  getHttrackProgress,
  waitForHttrackComplete,
} from "./httrack";
import { jwt } from "../constants.js";

// Skip long tests when running in watch mode.
const skipLongTests = process.env.npm_lifecycle_event === "test:watch";

describe("getHttrackArgs", () => {
  it("should return an array of arguments", async () => {
    const options = {
      url: new URL("https://intranet.justice.gov.uk/"),
      dest: "/tmp/test-snapshot",
      agency: "hq",
      jwt,
    };

    const args = getHttrackArgs(options);

    // Redacted args
    const redactedArgs = args.map((entry) => entry.replace(jwt, "***"));

    // Compare to snapshot
    expect(redactedArgs).toMatchSnapshot();
  });
});

describe("runHttrack", () => {
  beforeAll(() => {
    // Mock console.log so the tests are quiet.
    jest.spyOn(console, "log").mockImplementation(() => {});
  });

  afterAll(() => {
    // Restore console.log
    jest.restoreAllMocks();
  });

  it("should run httrack with arguments", async () => {
    const { listener, promise } = runHttrack([
      "http://www.all.net/",
      "-O",
      "/tmp/www.all.net",
      "+*.all.net/*",
      "-v",
      "-r2",
    ]);

    // listener should have a pid, it should be a number
    expect(listener.pid).toBeGreaterThan(0);

    const exitCode = await promise;

    expect(exitCode).toBe(0);
  });
});

describe("getHttrackProgress", () => {
  if (skipLongTests) {
    it.skip("should return progress", async () => {});
    return;
  }

  beforeAll(() => {
    // Mock console.log so the tests are quiet.
    jest.spyOn(console, "log").mockImplementation(() => {});
  });

  afterAll(() => {
    // Restore console.log
    jest.restoreAllMocks();
  });

  it("should return progress", async () => {
    // Delete the folder fontents /tmp/www.all.net
    fs.rmSync("/tmp/www.all.net", { recursive: true });

    // console.log(process.env);
    // Get some test data.
    const { listener, promise } = runHttrack([
      "http://www.all.net/",
      "-O",
      "/tmp/www.all.net",
      "+*.all.net/*",
      "-v",
      "-r10",
    ]);

    // await promise;

    // Wait for 0.5 seconds
    await new Promise((resolve) => setTimeout(resolve, 5000));

    /**
     * 1️⃣ Get the progress for the 1st time.
     */
    const results1 = await getHttrackProgress("/tmp/www.all.net");

    expect(results1.complete).toBe(false);
    expect(results1.rate).toBeGreaterThan(0);
    expect(results1.requestCount).toBeGreaterThan(0);

    /**
     * 2️⃣ Get the progress for the 2nd time - immediately after the process has finished.
     */

    await promise;

    const results2 = await getHttrackProgress("/tmp/www.all.net");

    expect(results2.complete).toBe(true);
    expect(results2.rate).toBeGreaterThan(0);
    expect(results2.requestCount).toBeGreaterThan(results1.requestCount);

    /**
     * 3️⃣ Get the progress for the 3rd time - 5 seconds after the process has finished.
     */

    await new Promise((resolve) => setTimeout(resolve, 5000));

    const results3 = await getHttrackProgress("/tmp/www.all.net");

    expect(results3.rate).toBe(0);
    expect(results3.complete).toBe(true);
  }, 20_000);
});

describe("waitForHttrackComplete", () => {
  beforeAll(() => {
    // Delete the folder fontents /tmp/www.all.net
    fs.rmSync("/tmp/www.all.net", { recursive: true });

    // Mock console.log so the tests are quiet.
    jest.spyOn(console, "log").mockImplementation(() => {});
  });

  afterAll(() => {
    // Restore console.log
    jest.restoreAllMocks();
  });

  it("should resolve when httrack is complete", async () => {
    const { promise } = runHttrack([
      "http://www.all.net/",
      "-O",
      "/tmp/www.all.net",
      "+*.all.net/*",
      "-v",
      "-r1",
    ]);

    const httrackComplete = await waitForHttrackComplete("/tmp/www.all.net");

    expect(httrackComplete).toStrictEqual({"timedOut": false});

    // Wait for the process to finish logging.
    await promise;
    await new Promise((resolve) => setTimeout(resolve, 500));
  });

  it("should resolve when timeout is reached", async () => {
    const { promise } = runHttrack([
      "http://www.all.net/",
      "-O",
      "/tmp/www.all.net",
      "+*.all.net/*",
      "-v",
      "-r5",
    ]);

    const httrackComplete = await waitForHttrackComplete("/tmp/www.all.net", 2);

    expect(httrackComplete).toStrictEqual({"timedOut": true});

    // Wait for the process to finish logging.
    await promise;
    await new Promise((resolve) => setTimeout(resolve, 500));
  }, 10_000);
});
