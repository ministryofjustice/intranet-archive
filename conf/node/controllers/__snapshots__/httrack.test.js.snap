// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getHttrackArgs should return an array of arguments 1`] = `
[
  "-%W",
  "/archiver/strip_x_amz_query_param.so",
  "https://intranet.justice.gov.uk/",
  "+*.png",
  "+*.gif",
  "+*.jpg",
  "+*.jpeg",
  "+*.css",
  "+*.js",
  "-ad.doubleclick.net/*",
  "-*intranet.justice.gov.uk/agency-switcher/",
  "-*intranet.justice.gov.uk/?*agency=*",
  "-*intranet.justice.gov.uk/?p=*",
  "-*intranet.justice.gov.uk/?page_id=*",
  "-*intranet.justice.gov.uk/wp-json/*/embed*",
  "-*intranet.justice.gov.uk/wp/*",
  "+*intranet.justice.gov.uk/?*agency=hq",
  "-s0",
  "-V",
  ""sed -i 's/srcset="[^"]*"//g' $0"",
  "-%k",
  "-F",
  "intranet-archive",
  "-%X",
  "Cookie: dw_agency=hq; jwt=***",
  "-O",
  "/tmp/test-snapshot",
]
`;
